### PR TITLE
Move IVsSolutionEvents to VisualStudioWorkspaceImpl

### PR DESCRIPTION
**Pull request by Pilchie**
In Lightweight solution load, we depend on
IVsSolutionEvents.OnAfterOpenSolution being called to populate the
workspace and create projects.  However we were subscribing from
VisualStudioProjectTracker, which we defer create until a project is
opened.  This meant we never got the event for the first time a solution
was opened, and so we failed to populate the workspace.

### CODE REVIEW

**Comment by Pilchie** [comment]
Tagging @dotnet/roslyn-ide for review.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 1]
```diff
@@ -506,12 +496,58 @@ internal void TryDisconnectExistingDeferredProject(IVsHierarchy hierarchy, strin
...
+            // check the token periodically, and whenever resuming from an "await"
+            _solutionParsingCancellationTokenSource.Cancel();
+            _solutionParsingCancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public void OnAfterCloseSolution()
+        {
+            AssertIsForeground();
+
+            if (VisualStudioWorkspaceImpl.IsDeferredSolutionLoadEnabled(_serviceProvider))
```
Comment (GearedToWar) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Do you want to store the value returned from this method in OnBeforeCloseSolution() and use it here? That way you have the correct value for the solution which just closed.
If the global and the solution-level setting of LSL are not the same, then the value returned here will be incorrect.

Please verify that scenario.

Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Thanks - updated and tested the matrix of solution vs global state.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 2]
```diff
@@ -1022,6 +1022,12 @@ protected override void Dispose(bool finalize)
             // workspace is going away. unregister this workspace from work coordinator
             StopSolutionCrawler();
 
+            // We should consider calling this here. It is commented out because Solution event tracking was 
+            // moved from VisualStudioProjectTracker, which is never Dispose()'d.  Rather than risk the 
+            // UnadviseSolutionEvents causing another issue (calling into dead COM objects, etc), we'll just 
+            // continue to skip it for now.
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: Note that `VisualStudioWorskspace.Dispose()` *is* currently called by MEF when it is shutting down.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 3]
```diff
@@ -506,12 +502,59 @@ internal void TryDisconnectExistingDeferredProject(IVsHierarchy hierarchy, strin
...
-            // NOTE: It is expected that the "as" will fail on Dev14, as IVsSolution7 was
-            // introduced in Dev15.  Be sure to handle the null result here.
-            var solution7 = _serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution7;
-            return solution7?.IsSolutionLoadDeferred() == true;
+            AssertIsForeground();
+
+            if (_deferredLoadWasEnabledForLastSolution)
+            {
+                // Copy to avoid modifying the collection while enumerating
+                var loadedProjects = ImmutableProjects.ToList();
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Yeap - Not going to touch in this PR.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl_IVsSolutionEvents.cs [hunk 4]
```diff
@@ -59,60 +80,23 @@ int IVsSolutionEvents.OnQueryCloseSolution(object pUnkReserved, ref int pfCancel
...
-
-            _solutionIsClosing = false;
-
+            DeferredState?.ProjectTracker.OnAfterCloseSolution();
             return VSConstants.S_OK;
         }
+
+        internal static bool IsDeferredSolutionLoadEnabled(IServiceProvider serviceProvider)
+        {
+            // NOTE: It is expected that the "as" will fail on Dev14, as IVsSolution7 was
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl_IVsSolutionEvents.cs]: It was, not going to touch in this PR.

Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl_IVsSolutionEvents.cs]: Note that when it's cleaned up, that should include removing the local interop definition, and adding the real reference, etc.

**Comment by Pilchie** [comment]
@jasonmalinowski Fixed the as cast, and working on getting an RPS run now.  Will fix the others in `master` after this merges.

**Pull request merged**